### PR TITLE
gh-297 Port security infrastructure to Angular4 stack

### PR DESCRIPTION
**Pull request by ghillert**
- Port the non-oauth-based security to Angular 4
- Add Roles directive

resolves https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/297

### CODE REVIEW

File: ui/src/app/auth/directives/roles.directive.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import { AfterViewInit, Directive, ElementRef, Input, Output, EventEmitter, HostListener,
+DoCheck, Renderer2 } from '@angular/core';
+
+import { AuthService } from '../auth.service';
+
+/**
+ * This directive is a helper for any situation where you need
```
Comment (jvalkeal) [inline ui/src/app/auth/directives/roles.directive.ts]: Did you copy this jsdoc from somewhere else? Looks a bit off.

File: ui/src/app/auth/directives/roles.directive.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    if (found) {
+      this.renderer.setStyle(this.elem.nativeElement, 'display', 'inherit');
+    } else {
+      this.renderer.setStyle(this.elem.nativeElement, 'display', 'none');
+    }
+  }
+  /**
+   * Initializes the state of the tri-state checkbox after component is initialized.
```
Comment (jvalkeal) [inline ui/src/app/auth/directives/roles.directive.ts]: Fix doc

**Comment by jvalkeal** [comment]
If I try this against plain server(no auth enabled), all navigation tabs are missing.

This is also showing secondary issues, If `apps list` is denied, tab is removed but actual content is still visible.

File: ui/src/app/shared/model/serialization/serializable.model.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export interface Serializable<T> {
+  deserialize(inputJson: Object): T;
+}
```
Comment (cppwfs) [inline ui/src/app/shared/model/serialization/serializable.model.ts]: carriage return at end of file.

File: ui/src/app/shared/model/index.ts [hunk 4]
```diff
@@ -4,3 +4,4 @@ export { ErrorHandler } from './error-handler';
 export { Expandable } from './expandable';
 export { Page } from './page';
 export { Selectable } from './selectable';
+export { Serializable  } from './serialization/serializable.model';
```
Comment (cppwfs) [inline ui/src/app/shared/model/index.ts]: carriage return at end of file.

File: ui/src/app/auth/login.component.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import { Component, OnInit } from '@angular/core';
```
Comment (cppwfs) [inline ui/src/app/auth/login.component.ts]: need to have a spec for this class.

File: ui/src/app/auth/logout.component.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import { Component, OnInit } from '@angular/core';
```
Comment (cppwfs) [inline ui/src/app/auth/logout.component.ts]: need to have a spec for this class.

File: ui/src/app/auth/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
+import { Injectable } from '@angular/core';
```
Comment (cppwfs) [inline ui/src/app/auth/auth.service.ts]: need to have a spec for this service.

**Comment by cppwfs** [comment]
The last commit resolved the issue where the Navigation tabs were not present.
Still having problem attempting to get into a secured server with the UI.  We may need add some info on how to do this in the readme.

File: ui/src/app/shared/model/index.ts (hunk 4 above)
Comment (ghillert) [inline ui/src/app/shared/model/index.ts]: Fixed

File: ui/src/app/shared/model/serialization/serializable.model.ts (hunk 3 above)
Comment (ghillert) [inline ui/src/app/shared/model/serialization/serializable.model.ts]: Fixed

**Comment by cppwfs** [comment]
I'm thinking the last commit was a progress checkpoint.   Still seeing the  Navigation tabs on the login page and some of the comments were not resolved.

File: ui/src/app/auth/support/security-aware-request-options.ts [hunk 8]
```diff
@@ -1,37 +1,75 @@
...
+
+  public get xAuthToken(): string {
+    return ValueHolder.xAuthToken;
+  }
+
+  merge(options: RequestOptions) {
     return super.merge(options);
   }
 }
+
```
Comment (cppwfs) [inline ui/src/app/auth/support/security-aware-request-options.ts]: ValueHolder has to be moved above DefaultRequestOptions.   Lint fun ;-)

**Comment by cppwfs** [comment]
Once security is activated, going to the http://localhost:4200 takes me to a page with just the header and a blank body.  From their I can click the profile button and select login.  It should take me to the login page.

**Comment by cppwfs** [comment]
Apps delete button still appears when 'VIEW' role is set.  It will take you to the `are you sure` modal.   At which time it will fail if you click yes.

**Comment by cppwfs** [comment]
Tasks definition button is still present when on the `tasks/apps`  'VIEW' role is set ( it should not be present).  It will take you to the create definition page, which will fail when you click the `submit` button.

File: ui/src/app/auth/roles-missing.component.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+ * @author Gunnar Hillert
+ */
+@Component({
+  template: `
+    <h1>Roles Missing</h1>
+    <p>
+      It appears that you are missing the proper roles to use the Dashboard. Please contact your administrator to rectify the
+      situation.
+    </p>
+  `,
```
Comment (cppwfs) [inline ui/src/app/auth/roles-missing.component.ts]: we can remove the comma

File: ui/src/app/auth/login.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    expect(passwordField.getAttribute('placeholder')).toBe('<Password>');
+
+    const loginButton: any = fixture.debugElement.query(By.css('#loginButton')).nativeElement;
+    expect(loginButton.innerHTML).toBe('Sign in');
+
+    const logoutButton: DebugElement = fixture.debugElement.query(By.css('#logoutButton'));
+    expect(logoutButton).toBe(null);
+    const backButton: DebugElement = fixture.debugElement.query(By.css('#logoutButton'));
+    expect(backButton).toBe(null);
+
```
Comment (cppwfs) [inline ui/src/app/auth/login.component.spec.ts]: An easy test to add  would be to verify that login & logout buttons work by verifying the navigation.   For example: https://github.com/spring-cloud/spring-cloud-dataflow-ui/blob/master/ui/src/app/about/about.component.spec.ts#L88

File: ui/src/app/tests/mocks/auth.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import { Observable } from 'rxjs/Observable';
+import { AppInfo } from '../../tasks/model/app-info';
```
Comment (cppwfs) [inline ui/src/app/tests/mocks/auth.ts]: AppInfo import can be removed it isn't used.

File: ui/src/app/tests/mocks/auth.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+import { Page} from '../../shared/model/page';
```
Comment (cppwfs) [inline ui/src/app/tests/mocks/auth.ts]: Page import can be removed it isn't used.

File: ui/src/app/tests/mocks/auth.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import { AppRegistration } from '../../shared/model/app-registration';
```
Comment (cppwfs) [inline ui/src/app/tests/mocks/auth.ts]: AppRegistration import can be removed it isn't used.

File: ui/src/app/tests/mocks/auth.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+import { TaskExecution } from '../../tasks/model/task-execution';
```
Comment (cppwfs) [inline ui/src/app/tests/mocks/auth.ts]: TaskExecution import can be removed it isn't used.

File: ui/src/app/tests/mocks/auth.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+ */
+export class MockAuthService {
+
+  public securityInfo = new SecurityInfo();
+
+  constructor() {
+  }
+
+  loadSecurityInfo(reconstituteSecurity = false): Observable<SecurityInfo> {
+    return Observable.of(this.securityInfo);
```
Comment (cppwfs) [inline ui/src/app/tests/mocks/auth.ts]: Is this method used?

**Comment by cppwfs** [comment]
Also found an issue.  I could be wrong.  When using role CREATE alone, I get the `Roles Missing` page.  Is role VIEW always required?

File: ui/src/app/auth/model/security-info.model.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
+import { Serializable } from '../../shared/model';
```
[the remark this answers is not in the document]
Comment (ghillert) [reply ui/src/app/auth/model/security-info.model.ts]: For the model classes, I created an `index.ts` file, so the import statements can be a bit shorter.

[the remark this answers is not in the document]
Comment (cppwfs) [reply ui/src/app/auth/model/security-info.model.ts]: ok.  But this is showing as a warning, until we correct it.

File: ui/src/app/auth/support/auth.guard.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+   * be redirected to the login url.
+   *
+   * @param route
+   * @param state
+   */
+  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot) {
+
+    const securityInfo = this.authService.securityInfo;
+    const routeNeedsAuthentication = route.data.authenticate;
+    const rolesNeeded: string[] = route.data.roles;
```
[the remark this answers is not in the document]
Comment (ghillert) [reply ui/src/app/auth/support/auth.guard.ts]: It is actually.

File: ui/src/app/auth/logout.component.ts (hunk 6 above)
Comment (ghillert) [reply ui/src/app/auth/logout.component.ts]: Done.

File: ui/src/app/auth/login.component.ts (hunk 5 above)
Comment (ghillert) [reply ui/src/app/auth/login.component.ts]: Done.

File: ui/src/app/auth/auth.service.ts (hunk 7 above)
Comment (ghillert) [reply ui/src/app/auth/auth.service.ts]: Done.

**Comment by ghillert** [comment]
@cppwfs Regarding `Also found an issue. I could be wrong. When using role CREATE alone, I get the Roles Missing page. Is role VIEW always required?`

Yes, the VIEW role is always required. The `Create` role does *not include* the view capability implicitly. Similar to the REST api behavior. Same behavior as in the old UI. CC @sabbyanandan

**Comment by ghillert** [comment]
Regarding `Apps delete button still appears when 'VIEW' role is set (it should not be present). It will take you to the are you sure modal. At which time it will fail if you click yes.`

Should be fixed now.

File: ui/src/app/auth/login.component.spec.ts (hunk 10 above)
Comment (ghillert) [reply ui/src/app/auth/login.component.spec.ts]: Create https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/378

Comment (cppwfs) [reply ui/src/app/auth/login.component.spec.ts]: sweet!

File: ui/src/app/auth/model/security-info.model.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (cppwfs) [reply ui/src/app/auth/model/security-info.model.ts]: Looks resolved.  sweet!

**Comment by cppwfs** [comment]
ng lint --type-check  shows `variable 'ValueHolder' used before declaration`.  Will fix on merge.

**Comment by cppwfs** [comment]
When role=VIEW, users can still access the app register, stream deploy, task create if they enter the url directly.  i.e.
apps/registration-apps
definitions/foo3/deploy
tasks/apps/AAA/create
Should I go ahead and add a story for this?

**Comment by cppwfs** [comment]
LGTM Opened  https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/379 to handle last issue.

**Comment by cppwfs** [comment]
Polished, squashed, merged.

**Pull request closed without merging**